### PR TITLE
Add ordinal and one-hot encodings

### LINKED ISSUE #14: Add one-hot preprocessing
A function to encode classes using one-hot, and possibly a reverse operation that just does argmax. This would make sense in the `Preprocessing` module.

### PR DESCRIPTION
**Pull request by msluszniak**
Closes #14.

### CODE REVIEW

File: lib/scholar/preprocessing.ex [hunk 1]
```diff
@@ -87,4 +87,75 @@ defmodule Scholar.Preprocessing do
     opts = keyword!(opts, threshold: 0, type: {:f, 32})
     (tensor > opts[:threshold]) |> Nx.as_type(opts[:type])
   end
+
+  @doc """
+  Encodes a tensor's values into integers from range 0 to `:num_classes - 1`
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:92]:
```suggestion
  Encodes a tensor's values into integers from range 0 to `:num_classes - 1`.
```

File: lib/scholar/preprocessing.ex [hunk 2]
```diff
@@ -87,4 +87,75 @@ defmodule Scholar.Preprocessing do
...
+  ## Examples
+
+      iex> Scholar.Preprocessing.ordinal_encoding(Nx.tensor([3, 2, 4, 56, 2, 4, 2]), num_classes: 4)
+      #Nx.Tensor<
+        s64[7]
+        [1, 0, 2, 3, 0, 2, 0]
+      >
+  """
+
+  @spec ordinal_encoding(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:107]:
```suggestion
  """
  @spec ordinal_encoding(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
```

File: lib/scholar/preprocessing.ex [hunk 3]
```diff
@@ -87,4 +87,75 @@ defmodule Scholar.Preprocessing do
...
+      ])
+      |> Nx.argsort(direction: :desc)
+      |> then(fn x -> x[[0..(num_classes - 1)]] end)
+
+    unique_values = Nx.take(sorted, representative_mask)
+
+    Nx.equal(
+      Nx.reshape(tensor, {num_samples, 1}),
+      Nx.broadcast(unique_values, {num_samples, num_classes})
+    )
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:128]:
```suggestion
    # A mask with a single 1 in every group of equal values,
    representative_mask =
      Nx.concatenate([
        Nx.not_equal(sorted[0..-2//1], sorted[1..-1//1]),
        Nx.tensor([1])
      ])
      
    representative_indices =
      representative_mask
      |> Nx.argsort(direction: :desc)
      |> Nx.slice_along_axis(0, num_classes)

    representative_values = Nx.take(sorted, representative_indices)

    Nx.equal(
      Nx.reshape(tensor, {num_samples, 1}),
      Nx.reshape(representative_values, {1, num_classes})
    )
```

File: lib/scholar/preprocessing.ex [hunk 4]
```diff
@@ -87,4 +87,75 @@ defmodule Scholar.Preprocessing do
...
+          [0, 0, 0, 1],
+          [0, 0, 1, 0],
+          [0, 1, 0, 0],
+          [0, 1, 0, 0],
+          [1, 0, 0, 0]
+        ]
+      >
+  """
+
+  @spec one_hot_encoding(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:157]:
```suggestion
  """
  @spec one_hot_encoding(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
```

File: lib/scholar/preprocessing.ex [hunk 5]
```diff
@@ -87,4 +87,76 @@ defmodule Scholar.Preprocessing do
...
+      >
+  """
+  @spec ordinal_encoding(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
+  defn ordinal_encoding(tensor, opts \\ []) do
+    {num_samples} = Nx.shape(tensor)
+    opts = keyword!(opts, [:num_classes])
+    sorted = Nx.sort(tensor)
+    num_classes = opts[:num_classes]
+
+    # A mask with a single 1 in every group of equal values,
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:113]:
```suggestion
    # A mask with a single 1 in every group of equal values
```

File: lib/scholar/preprocessing.ex [hunk 6]
```diff
@@ -87,4 +87,76 @@ defmodule Scholar.Preprocessing do
...
+    Nx.equal(
+      Nx.reshape(tensor, {num_samples, 1}),
+      Nx.reshape(representative_values, {1, num_classes})
+    )
+    |> Nx.argmax(axis: 1)
+  end
+
+  @doc """
+  Encode labels as a one-hot numeric tensor. Labels must be integers from 0 to `:num_classes - 1`.
+  If the data does not meet the condition, please use ordinal_encoding first.
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:136]:
```suggestion
  Encode labels as a one-hot numeric tensor.
  
  Labels must be integers from 0 to `:num_classes - 1`. If the data does
  not meet the condition, please use `ordinal_encoding/2` first.
```

File: lib/scholar/preprocessing.ex [hunk 7]
```diff
@@ -87,4 +87,78 @@ defmodule Scholar.Preprocessing do
...
+  ## Examples
+
+      iex> Scholar.Preprocessing.ordinal_encoding(Nx.tensor([3, 2, 4, 56, 2, 4, 2]), num_classes: 4)
+      #Nx.Tensor<
+        s64[7]
+        [1, 0, 2, 3, 0, 2, 0]
+      >
+  """
+  @spec ordinal_encoding(tensor :: Nx.Tensor.t(), opts :: Keyword.t()) :: Nx.Tensor.t()
+  defn ordinal_encoding(tensor, opts \\ []) do
```
Comment (jonatanklosko) [inline lib/scholar/preprocessing.ex:107]: @josevalim @polvalente @seanmor5 any preferences on the naming? Perhaps `encode_oridinal`/`encode_one_hot`?

Comment (josevalim) [reply lib/scholar/preprocessing.ex:107]: `one_hot_encode`

Comment (josevalim) [reply lib/scholar/preprocessing.ex:107]: We could also call it `one_hot_encoder`, which would match `standard_scaler`. But perhaps it should be `one_hot_encode` and `standard_scale`.

Comment (jonatanklosko) [reply lib/scholar/preprocessing.ex:107]: +1 for the `one_hot_encode` and `standard_scale`.

Comment (josevalim) [reply lib/scholar/preprocessing.ex:107]: @msluszniak let's go with one_hot_encode then! I will open up an issue to rename the other.

**Review by jonatanklosko — APPROVED** [review]

**Pull request merged**